File: LensConnect_SampleProgram_Linux_C_x86_x64_1.0.0/LensConnect_Controller/C/LensCtrl.c
#include <time.h>	// for nanosleep
#include "LensCtrl.h"

#define MILLI_SEC 1000000L

uint16_t zoomMaxAddr, zoomMinAddr, focusMaxAddr, focusMinAddr;
uint16_t irisMaxAddr, irisMinAddr, optFilMaxAddr;
uint16_t zoomCurrentAddr, focusCurrentAddr, irisCurrentAddr, optCurrentAddr;
uint16_t zoomSpeedPPS, focusSpeedPPS, irisSpeedPPS;
uint16_t status2;

void MsSleep(int n) {
	struct timespec ts;
	ts.tv_sec  = 0;
	ts.tv_nsec = MILLI_SEC;
	nanosleep(&ts, NULL);
}

uint16_t NoErrChk2BytesRead(uint16_t segmentOffset) {
	UsbRead(segmentOffset, 2);


	return UsbRead2Bytes();
}
int ModelName(char* model) {
	int retval = UsbRead(LENS_MODEL_NAME, LENSMODEL_LENGTH);
	strncpy(model, (const char*)receivedData, (LENSMODEL_LENGTH ));
	model[LENSMODEL_LENGTH] ='\0';
	return retval;
}
int UserAreaRead(char* userName) {
	int retval =  UsbRead(USER_AREA, USERAREA_LENGTH);
	strncpy(userName, (const char*)receivedData, (USERAREA_LENGTH));
	userName[USERAREA_LENGTH] = '\0';
	return retval;
}

int FWVersion(char* version) {
	int retval = UsbRead(FIRMWARE_VERSION, DATA_LENGTH);
	if ((retval == HID_SMBUS_SUCCESS) & (receivedData != NULL)) {
		sprintf(version, "%d.%02d", receivedData[0], receivedData[1]);
	}
	else {
		version[0] = '\0';
	}
	return retval;
}
int ProtocolVersion(char* version) {
	int retval = UsbRead(PROTOCOL_VERSION, DATA_LENGTH);
	if ((retval == HID_SMBUS_SUCCESS) & (receivedData != NULL)) {
		sprintf(version, "%d.%02d", receivedData[0], receivedData[1]);
	}
	else {
		version[0] = '\0';
	}
	return retval;
}

int LensRevision(int* revision) {
	int retval = UsbRead(LENS_REVISION, DATA_LENGTH);
	*revision = UsbRead2Bytes();
	return retval;
}

int LensAddress(int* i2cAddress) {
	int retval = UsbRead(LENS_ADDRESS, LENSADDRESS_LENGTH);
	*i2cAddress = receivedData[0];
	return retval;
}
int CapabilitiesRead(uint16_t* capabilities) {
	int retval = UsbRead(CAPABILITIES, DATA_LENGTH);
	*capabilities = UsbRead2Bytes();
	return retval;
}
int Status1Read(uint16_t* status1) {
	int retval = UsbRead(STATUS1, DATA_LENGTH);
	*status1 = UsbRead2Bytes();
	return retval;
}
int Status2ReadSet() {
	int retval = UsbRead(STATUS2, DATA_LENGTH);
	if (retval != HID_SMBUS_SUCCESS)
		return retval;
	status2 = UsbRead2Bytes();
	return retval;
}
int TempKelvin(int* kelvinValue) {
	int retval = UsbRead(TEMPERATURE_VAL, DATA_LENGTH);
	*kelvinValue = UsbRead2Bytes();
	return retval;
}

int UserAreaWrite(char* userName) {
	char sendData[USERAREA_LENGTH + SEGMENTOFFSET_LENGTH + 1];
	sendData[0] = USER_AREA >> 8;
	sendData[1] = USER_AREA;
	strncpy(sendData + SEGMENTOFFSET_LENGTH, userName, USERAREA_LENGTH + 1);
	int retval = HidSmbus_WriteRequest(connectedDevice, i2cAddr, 
	(unsigned char*)sendData, USERAREA_LENGTH + SEGMENTOFFSET_LENGTH);
	return retval;
}
int WaitCalc(uint16_t moveValue, int speedPPS) {
	int waitTime = WAIT_MAG * moveValue / speedPPS;
	if (MINIMUM_WAIT > waitTime)
		waitTime = MINIMUM_WAIT;
	return waitTime;
}

int ZoomCurrentAddrReadSet() {
	int retval = UsbRead(ZOOM_POSITION_VAL, DATA_LENGTH);
	zoomCurrentAddr = UsbRead2Bytes();
	return retval;
}
int ZoomParameterReadSet() {
	int retval = UsbRead(ZOOM_POSITION_MIN, DATA_LENGTH);
	if (retval != HID_SMBUS_SUCCESS)
		return retval;
	zoomMinAddr = UsbRead2Bytes();

	retval = UsbRead(ZOOM_POSITION_MAX, DATA_LENGTH);
	if (retval != HID_SMBUS_SUCCESS)
		return retval;
	zoomMaxAddr = UsbRead2Bytes();

	retval = UsbRead(ZOOM_SPEED_VAL, DATA_LENGTH);
	if (retval != HID_SMBUS_SUCCESS)
		return retval;
	zoomSpeedPPS = UsbRead2Bytes();

	return retval;
}
int ZoomBacklashRead(uint16_t* flag) {
	int retval = UsbRead(ZOOM_BACKLASH_CANCEL, DATA_LENGTH);
	*flag = UsbRead2Bytes();
	return retval;
}
int ZoomBacklashWrite(uint16_t flag) {
	int retval = UsbWrite(ZOOM_BACKLASH_CANCEL, flag);
	return retval;
}
int ZoomSpeedMinRead(uint16_t* speedPPS) {
	int retval = UsbRead(ZOOM_SPEED_MIN, DATA_LENGTH);
	*speedPPS = UsbRead2Bytes();
	return retval;
}
int ZoomSpeedMaxRead(uint16_t* speedPPS) {
	int retval = UsbRead(ZOOM_SPEED_MAX, DATA_LENGTH);
	*speedPPS = UsbRead2Bytes();
	return retval;
}
int ZoomSpeedWrite(uint16_t speedPPS) {
	int retval = UsbWrite(ZOOM_SPEED_VAL, speedPPS);
	if (retval != HID_SMBUS_SUCCESS)
		return retval;
	MsSleep(1);
	retval = UsbRead(ZOOM_SPEED_VAL, DATA_LENGTH);
	if (retval != HID_SMBUS_SUCCESS)
		return retval;
	zoomSpeedPPS = UsbRead2Bytes();
	return retval;
}
int ZoomCountValRead(uint32_t* count) {
	int retval = UsbRead(ZOOM_COUNT_VAL, LENSCOUNT_LENGTH);
	*count = CountRead();
	return retval;
}
int ZoomCountMaxRead(uint32_t* count) {
	int retval = UsbRead(ZOOM_COUNT_MAX, LENSCOUNT_LENGTH);
	*count = CountRead();
	return retval;
}
int ZoomInit() {
	int waitTime = WaitCalc((zoomMaxAddr - zoomMinAddr), zoomSpeedPPS);
	int retval = UsbWrite(ZOOM_INITIALIZE, INIT_RUN_BIT);
	if (retval == HID_SMBUS_SUCCESS) {
		retval = StatusWait(STATUS1, ZOOM_MASK, waitTime);
		if (retval == HID_SMBUS_SUCCESS) {
			retval = UsbRead(ZOOM_POSITION_VAL, DATA_LENGTH);
			if (retval == HID_SMBUS_SUCCESS) {
				zoomCurrentAddr = UsbRead2Bytes();
				Status2ReadSet();
				return retval;
			}
		}
	}
	return retval;
}
int ZoomMove(uint16_t addrData) {
	uint16_t moveVal = abs(addrData - zoomCurrentAddr);
	int waitTime = WaitCalc(moveVal, zoomSpeedPPS);
	int retval = DeviceMove(ZOOM_POSITION_VAL, &addrData, ZOOM_MASK, waitTime);
	if (retval == HID_SMBUS_SUCCESS)
		zoomCurrentAddr = addrData;
	return retval;
}

int FocusCurrentAddrReadSet() {
	int retval = UsbRead(FOCUS_POSITION_VAL, DATA_LENGTH);
	focusCurrentAddr = UsbRead2Bytes();
	return retval;
}
int FocusParameterReadSet() {
	int retval = UsbRead(FOCUS_MECH_STEP_MIN, DATA_LENGTH);
	if (retval != HID_SMBUS_SUCCESS)
		return retval;
	focusMinAddr = UsbRead2Bytes();

	retval = UsbRead(FOCUS_MECH_STEP_MAX, DATA_LENGTH);
	if (retval != HID_SMBUS_SUCCESS)
		return retval;
	focusMaxAddr = UsbRead2Bytes();

	retval = UsbRead(FOCUS_SPEED_VAL, DATA_LENGTH);
	if (retval != HID_SMBUS_SUCCESS)
		return retval;
	focusSpeedPPS = UsbRead2Bytes();

	return retval;
}
int FocusBacklashRead(uint16_t* flag) {
	int retval = UsbRead(FOCUS_BACKLASH_CANCEL, DATA_LENGTH);
	*flag = UsbRead2Bytes();
	return retval;
}
int FocusBacklashWrite(uint16_t flag) {
	int retval = UsbWrite(FOCUS_BACKLASH_CANCEL, flag);
	return retval;
}
int FocusSpeedMinRead(uint16_t* speedPPS) {
	int retval = UsbRead(FOCUS_SPEED_MIN, DATA_LENGTH);
	*speedPPS = UsbRead2Bytes();
	return retval;
}
int FocusSpeedMaxRead(uint16_t* speedPPS) {
	int retval = UsbRead(FOCUS_SPEED_MAX, DATA_LENGTH);
	*speedPPS = UsbRead2Bytes();
	return retval;
}
int FocusSpeedWrite(uint16_t speedPPS) {
	int retval = UsbWrite(FOCUS_SPEED_VAL, speedPPS);
	if (retval != HID_SMBUS_SUCCESS)
		return retval;
	MsSleep(1);
	retval = UsbRead(FOCUS_SPEED_VAL, DATA_LENGTH);
	if (retval != HID_SMBUS_SUCCESS)
		return retval;
	focusSpeedPPS = UsbRead2Bytes();
	return retval;
}
int FocusCountValRead(uint32_t* count) {
	int retval = UsbRead(FOCUS_COUNT_VAL, LENSCOUNT_LENGTH);
	*count = CountRead();
	return retval;
}
int FocusCountMaxRead(uint32_t* count) {
	int retval = UsbRead(FOCUS_COUNT_MAX, LENSCOUNT_LENGTH);
	*count = CountRead();
	return retval;
}
int FocusInit() {
	int waitTime = WaitCalc((focusMaxAddr - focusMinAddr), focusSpeedPPS);
	int retval = UsbWrite(FOCUS_INITIALIZE, INIT_RUN_BIT);
	if (retval == HID_SMBUS_SUCCESS) {
		retval = StatusWait(STATUS1, FOCUS_MASK, waitTime);
		if (retval == HID_SMBUS_SUCCESS) {
			retval = UsbRead(FOCUS_POSITION_VAL, DATA_LENGTH);
			if (retval == HID_SMBUS_SUCCESS) {
				focusCurrentAddr = UsbRead2Bytes();
				Status2ReadSet();
				return retval;
			}
		}
	}
	return retval;
}
int FocusMove(uint16_t addrData) {
	uint16_t moveVal = abs(addrData - focusCurrentAddr);
	int waitTime = WaitCalc(moveVal, focusSpeedPPS);
	int retval = DeviceMove(FOCUS_POSITION_VAL, &addrData, FOCUS_MASK, waitTime);
	if (retval == HID_SMBUS_SUCCESS)
		focusCurrentAddr = addrData;
	return retval;
}

int IrisCurrentAddrReadSet() {
	int retval = UsbRead(IRIS_POSITION_VAL, DATA_LENGTH);
	irisCurrentAddr = UsbRead2Bytes();
	return retval;
}
int IrisParameterReadSet() {
	int retval = UsbRead(IRIS_MECH_STEP_MIN, DATA_LENGTH);
	if (retval != HID_SMBUS_SUCCESS)
		return retval;
	irisMinAddr = UsbRead2Bytes();

	retval = UsbRead(IRIS_MECH_STEP_MAX, DATA_LENGTH);
	if (retval != HID_SMBUS_SUCCESS)
		return retval;
	irisMaxAddr = UsbRead2Bytes();

	retval = UsbRead(IRIS_SPEED_VAL, DATA_LENGTH);
	if (retval != HID_SMBUS_SUCCESS)
		return retval;
	irisSpeedPPS = UsbRead2Bytes();

	return retval;
}
int IrisBacklashRead(uint16_t* flag) {
	int retval = UsbRead(IRIS_BACKLASH_CANCEL, DATA_LENGTH);
	*flag = UsbRead2Bytes();
	return retval;
}
int IrisBacklashWrite(uint16_t flag) {
	int retval = UsbWrite(IRIS_BACKLASH_CANCEL, flag);
	return retval;
}
int IrisSpeedMinRead(uint16_t* speedPPS) {
	int retval = UsbRead(IRIS_SPEED_MIN, DATA_LENGTH);
	*speedPPS = UsbRead2Bytes();
	return retval;
}
int IrisSpeedMaxRead(uint16_t* speedPPS) {
	int retval = UsbRead(IRIS_SPEED_MAX, DATA_LENGTH);
	*speedPPS = UsbRead2Bytes();
	return retval;
}
int IrisSpeedWrite(uint16_t speedPPS) {
	int retval = UsbWrite(IRIS_SPEED_VAL, speedPPS);
	if (retval != HID_SMBUS_SUCCESS)
		return retval;
	MsSleep(1);
	retval = UsbRead(IRIS_SPEED_VAL, DATA_LENGTH);
	if (retval != HID_SMBUS_SUCCESS)
		return retval;
	irisSpeedPPS = UsbRead2Bytes();
	return retval;
}
int IrisCountValRead(uint32_t* count) {
	int retval = UsbRead(IRIS_COUNT_VAL, LENSCOUNT_LENGTH);
	*count = CountRead();
	return retval;
}
int IrisCountMaxRead(uint32_t* count) {
	int retval = UsbRead(IRIS_COUNT_MAX, LENSCOUNT_LENGTH);
	*count = CountRead();
	return retval;
}
int IrisInit() {
	int waitTime = WaitCalc((irisMaxAddr - irisMinAddr), irisSpeedPPS);
	int retval = UsbWrite(IRIS_INITIALIZE, INIT_RUN_BIT);
	if (retval == HID_SMBUS_SUCCESS) {
		retval = StatusWait(STATUS1, IRIS_MASK, waitTime);
		if (retval == HID_SMBUS_SUCCESS) {
			retval = UsbRead(IRIS_POSITION_VAL, DATA_LENGTH);
			if (retval == HID_SMBUS_SUCCESS) {
				irisCurrentAddr = UsbRead2Bytes();
				Status2ReadSet();
				return retval;
			}
		}
	}
	return retval;
}
int IrisMove(uint16_t addrData) {
	uint16_t moveVal = abs(addrData - irisCurrentAddr);
	int waitTime = WaitCalc(moveVal, irisSpeedPPS);
	int retval = DeviceMove(IRIS_POSITION_VAL, &addrData, IRIS_MASK, waitTime);
	if (retval == HID_SMBUS_SUCCESS)
		irisCurrentAddr = addrData;
	return retval;
}

int OptFilterCurrentAddrReadSet() {
	int retval = UsbRead(OPT_FILTER_POSITION_VAL, DATA_LENGTH);
	optCurrentAddr = UsbRead2Bytes();
	return retval;
}
int OptFilterParameterReadSet() {
	int retval = UsbRead(OPT_FILTER_MECH_STEP_MAX, DATA_LENGTH);
	if (retval != HID_SMBUS_SUCCESS)
		return retval;
	optFilMaxAddr = UsbRead2Bytes();

	retval = OptFilterCurrentAddrReadSet();
	return retval;
}
int OptFilterCountValRead(uint32_t* count) {
	int retval = UsbRead(OPT_FILTER_COUNT_VAL, LENSCOUNT_LENGTH);
	*count = CountRead();
	return retval;
}
int OptFilterCountMaxRead(uint32_t* count) {
	int retval = UsbRead(OPT_FILTER_COUNT_MAX, LENSCOUNT_LENGTH);
	*count = CountRead();
	return retval;
}
int OptFilterInit() {
	int waitTime = WaitCalc((optFilMaxAddr + 1), OPT_FILTER_SPEED);
	int retval = UsbWrite(OPT_FILTER_INITIALIZE, INIT_RUN_BIT);
	if (retval == HID_SMBUS_SUCCESS) {
		retval = StatusWait(STATUS1, OPT_FILTER_MASK, waitTime);
		if (retval == HID_SMBUS_SUCCESS) {
			retval = UsbRead(OPT_FILTER_POSITION_VAL, DATA_LENGTH);
			if (retval == HID_SMBUS_SUCCESS) {
				optCurrentAddr = UsbRead2Bytes();
				Status2ReadSet();
				return retval;
			}
		}
	}
	return retval;
}
int OptFilterMove(uint16_t addrData) {
	uint16_t moveVal = abs(addrData - optCurrentAddr);
	int waitTime = WaitCalc((moveVal + 1), OPT_FILTER_SPEED);
	int retval = DeviceMove( OPT_FILTER_POSITION_VAL, &addrData, OPT_FILTER_MASK, waitTime);
	if (retval == HID_SMBUS_SUCCESS)
		optCurrentAddr = addrData;
	return retval;
}
int DeviceMove(uint16_t segmentOffset, uint16_t *addrData, uint16_t mask , int waitTime) {
	int retval = UsbWrite(segmentOffset, *addrData);
	if (retval == HID_SMBUS_SUCCESS) {
		retval = StatusWait(STATUS1, mask, waitTime);
		if (retval == HID_SMBUS_SUCCESS) {
			retval = UsbRead(segmentOffset, DATA_LENGTH);
			if (retval != HID_SMBUS_SUCCESS)
				return retval;
			*addrData = UsbRead2Bytes();
			return retval;
		}
		return retval;
	}
	return retval;
}
int StatusWait(uint16_t segmentOffset, uint16_t statusMask, int waitTime) {
	int tmp = 0;
	uint16_t readStatus;
	int retval;
	do {
		retval = UsbRead(segmentOffset, DATA_LENGTH);
		if (retval != HID_SMBUS_SUCCESS)
			return retval;

		readStatus = UsbRead2Bytes();
		tmp += 1;
		if (tmp >= LOW_HIGH_WAIT)
			return LOWHI_ERROR;

	} while ((readStatus & statusMask) != statusMask);

	tmp = 0;
	do {
		retval = UsbRead(segmentOffset, DATA_LENGTH);
		if (retval != HID_SMBUS_SUCCESS)
			return retval;

		readStatus = UsbRead2Bytes();
		tmp += 1;
		if (tmp >= waitTime)
			return HILOW_ERROR;
		
		MsSleep(1);
	} while ((readStatus & statusMask) != 0);

	return retval;
}
const char *ErrorTxt(int retval) {
	switch (retval) {
	case 0x01:
		return "Device not found.\n";
	case 0x02:
		return "Invalid handle.\n";
	case 0x03:
		return "Invalid device object.\n";
	case 0x04:
		return "Invalid parameter.\n";
	case 0x05:
		return "Invalid request length.\n";
	case 0x10:
		return "Read error.\n";
	case 0x11:
		return "Write error.\n";
	case 0x12:
		return "Read time out.\n";
	case 0x13:
		return "Write time out.\n";
	case 0x14:
		return "Device IO failed.\n";
	case 0x15:
		return "Device access error.\r\nThe device may already be running.\n";
	case 0x16:
		return "Device not supported.\n";
	case 0x50:
		return "StatusBit hi error.\n";
	case 0x51:
		return "StatusBit low error.\n";
	}
	return "";
}

File: LensCtrl.py
from SLABHIDDevice import HID_DEVICE_SUCCESS
import DevAddr as DA
import ConfigVal as CV
import UsbCtrl
import time

zoomMaxAddr = 0
zoomMinAddr = 0
zoomCurrentAddr = 0
zoomSpeedPPS = 0

focusMaxAddr = 0
focusMinAddr = 0
focusCurrentAddr = 0
focusSpeedPPS = 0

irisMaxAddr = 0
irisMinAddr = 0
irisCurrentAddr = 0
irisSpeedPPS = 0

optFilMaxAddr = 0
optCurrentAddr = 0
status2 = 0

def NoErrChk2BytesIntRead(segmentOffset):
    retval, IntData = UsbCtrl.UsbRead2BytesInt(segmentOffset)
    return IntData

def ModelName():
    retval, data = UsbCtrl.UsbRead(DA.LENS_MODEL_NAME, CV.LENSMODEL_LENGTH)
    model = str(data,'utf-8')
    return retval, model

def UserAreaRead():
    retval, data = UsbCtrl.UsbRead(DA.USER_AREA, CV.USERAREA_LENGTH)
    userName = str(data,'utf-8')
    return retval, userName

def FWVersion():
    retval, data = UsbCtrl.UsbRead(DA.FIRMWARE_VERSION, CV.DATA_LENGTH)
    version = str(data[0]) + "." + format(data[1], '02d')
    return retval, version

def ProtocolVersion():
    retval, data = UsbCtrl.UsbRead(DA.PROTOCOL_VERSION, CV.DATA_LENGTH)
    version = str(data[0]) + "." + format(data[1], '02d')
    return retval, version

def LensRevision():
    retval, data = UsbCtrl.UsbRead2BytesInt(DA.LENS_REVISION)
    return retval, data

def LensAddress():
    retval, data = UsbCtrl.UsbRead(DA.LENS_ADDRESS, CV.DATA_LENGTH)
    i2cAddress = int(data[0])
    return retval, i2cAddress

def CapabilitiesRead():
    retval, data = UsbCtrl.UsbRead2BytesInt(DA.CAPABILITIES)
    return retval, data

def Status1Read():
    retval, data = UsbCtrl.UsbRead2BytesInt(DA.STATUS1)
    return retval, data

def Status2ReadSet():
    global status2
    retval, status2 = UsbCtrl.UsbRead2BytesInt(DA.STATUS2)
    return retval, status2

def TempKelvinVal():
    retval, data = UsbCtrl.UsbRead2BytesInt(DA.TEMPURATURE_VAL)
    return retval, data

def TempKelvinMin():
    retval, data = UsbCtrl.UsbRead2BytesInt(DA.TEMPURATURE_MIN)
    return retval, data

def TempKelvinMax():
    retval, data = UsbCtrl.UsbRead2BytesInt(DA.TEMPURATURE_MAX)
    return retval, data

def UserAreaWrite(UserName):
    UserNameSize = len(UserName)
    if (UserNameSize < CV.USERAREA_LENGTH):
        space = CV.USERAREA_LENGTH -UserNameSize
        for num in range(space):
            UserName = UserName + " "
            num += 1
    UserNameSize = len(UserName)
    retval = UsbCtrl.UsbWriteString(DA.USER_AREA, UserName)
    return retval

def WaitCalc(moveValue, speedPPS):
    waitTime = CV.WAIT_MAG *moveValue / speedPPS
    if (CV.MINIMUM_WAIT > waitTime):
        waitTime = CV.MINIMUM_WAIT
    return waitTime

def ZoomCurrentAddrReadSet():
    global zoomCurrentAddr
    retval, zoomCurrentAddr = UsbCtrl.UsbRead2BytesInt(DA.ZOOM_POSITION_VAL)
    return retval

def ZoomParameterReadSet():
    global zoomMinAddr
    global zoomMaxAddr
    global zoomSpeedPPS
    retval, zoomMinAddr = UsbCtrl.UsbRead2BytesInt(DA.ZOOM_POSITION_MIN)
    if (retval != HID_DEVICE_SUCCESS):
        return retval
    retval, zoomMaxAddr = UsbCtrl.UsbRead2BytesInt(DA.ZOOM_POSITION_MAX)
    if (retval != HID_DEVICE_SUCCESS):
        return retval
    retval, zoomSpeedPPS = UsbCtrl.UsbRead2BytesInt(DA.ZOOM_SPEED_VAL)
    if (retval != HID_DEVICE_SUCCESS):
        return retval
    return retval

def ZoomBacklashRead():
    retval, data = UsbCtrl.UsbRead2BytesInt(DA.ZOOM_BACKLASH_CANCEL)
    return retval, data

def ZoomBacklashWrite(flag):
    retval = UsbCtrl.UsbWrite(DA.ZOOM_BACKLASH_CANCEL, flag)
    return retval

def ZoomSpeedMinRead():
    retval, data = UsbCtrl.UsbRead2BytesInt(DA.ZOOM_SPEED_MIN)
    return retval, data

def ZoomSpeedMaxRead():
    retval, data = UsbCtrl.UsbRead2BytesInt(DA.ZOOM_SPEED_MAX)
    return retval, data

def ZoomSpeedWrite(speedPPS):
    global zoomSpeedPPS
    retval = UsbCtrl.UsbWrite(DA.ZOOM_SPEED_VAL, speedPPS)
    if (retval != HID_DEVICE_SUCCESS):
        return retval
    time.sleep(0.001)
    retval, zoomSpeedPPS = UsbCtrl.UsbRead2BytesInt(DA.ZOOM_SPEED_VAL)
    return retval

def ZoomCountValRead():
    retval, count = UsbCtrl.CountRead(DA.ZOOM_COUNT_VAL)
    return retval, count

def ZoomCountMaxRead():
    retval, count = UsbCtrl.CountRead(DA.ZOOM_COUNT_MAX)
    return retval, count

def ZoomInit():
    global zoomMaxAddr
    global zoomMinAddr
    global zoomSpeedPPS
    global zoomCurrentAddr
    waitTime = WaitCalc((zoomMaxAddr - zoomMinAddr), zoomSpeedPPS)
    retval = UsbCtrl.UsbWrite(DA.ZOOM_INITIALIZE, CV.INIT_RUN_BIT)
    if (retval == HID_DEVICE_SUCCESS):
        retval = StatusWait(DA.STATUS1, CV.ZOOM_MASK, waitTime)
        if (retval == HID_DEVICE_SUCCESS):
            retval, zoomCurrentAddr = UsbCtrl.UsbRead2BytesInt(DA.ZOOM_POSITION_VAL)
            if (retval == HID_DEVICE_SUCCESS):
                Status2ReadSet()
                return retval
    return retval

def ZoomMove(addrData):
    global zoomCurrentAddr
    global zoomSpeedPPS
    moveVal = abs(addrData - zoomCurrentAddr)
    waitTime = WaitCalc(moveVal, zoomSpeedPPS)
    retval, data = DeviceMove(DA.ZOOM_POSITION_VAL, addrData, CV.ZOOM_MASK, waitTime)
    if (retval == HID_DEVICE_SUCCESS):
        zoomCurrentAddr = data
    return retval

def FocusCurrentAddrReadSet():
    global focusCurrentAddr
    retval, focusCurrentAddr = UsbCtrl.UsbRead2BytesInt(DA.FOCUS_POSITION_VAL)
    return retval

def FocusParameterReadSet():
    global focusMinAddr
    global focusMaxAddr
    global focusSpeedPPS
    retval, focusMinAddr = UsbCtrl.UsbRead2BytesInt(DA.FOCUS_MECH_STEP_MIN)
    if (retval != HID_DEVICE_SUCCESS):
        return retval
    retval, focusMaxAddr = UsbCtrl.UsbRead2BytesInt(DA.FOCUS_MECH_STEP_MAX)
    if (retval != HID_DEVICE_SUCCESS):
        return retval
    retval, focusSpeedPPS = UsbCtrl.UsbRead2BytesInt(DA.FOCUS_SPEED_VAL)
    if (retval != HID_DEVICE_SUCCESS):
        return retval
    return retval

def FocusBacklashRead():
    retval, data = UsbCtrl.UsbRead2BytesInt(DA.FOCUS_BACKLASH_CANCEL)
    return retval, data
    
def FocusBacklashWrite(flag):
    retval = UsbCtrl.UsbWrite(DA.FOCUS_BACKLASH_CANCEL, flag)
    return retval

def FocusSpeedMinRead():
    retval, data = UsbCtrl.UsbRead2BytesInt(DA.FOCUS_SPEED_MIN)
    return retval, data

def FocusSpeedMaxRead():
    retval, data = UsbCtrl.UsbRead2BytesInt(DA.FOCUS_SPEED_MAX)
    return retval, data

def FocusSpeedWrite(speedPPS):
    global focusSpeedPPS
    retval = UsbCtrl.UsbWrite(DA.FOCUS_SPEED_VAL, speedPPS)
    if (retval != HID_DEVICE_SUCCESS):
        return retval
    time.sleep(0.001)
    retval, focusSpeedPPS = UsbCtrl.UsbRead2BytesInt(DA.FOCUS_SPEED_VAL)
    return retval

def FocusCountValRead():
    retval, count = UsbCtrl.CountRead(DA.FOCUS_COUNT_VAL)
    return retval, count

def FocusCountMaxRead():
    retval, count = UsbCtrl.CountRead(DA.FOCUS_COUNT_MAX)
    return retval, count

def FocusInit():
    global focusMaxAddr
    global focusMinAddr
    global focusSpeedPPS
    global focusCurrentAddr
    waitTime = WaitCalc((focusMaxAddr - focusMinAddr), focusSpeedPPS)
    retval = UsbCtrl.UsbWrite(DA.FOCUS_INITIALIZE, CV.INIT_RUN_BIT)
    if (retval == HID_DEVICE_SUCCESS):
        retval = StatusWait(DA.STATUS1, CV.FOCUS_MASK, waitTime)
        if (retval == HID_DEVICE_SUCCESS):
            retval, focusCurrentAddr = UsbCtrl.UsbRead2BytesInt(DA.FOCUS_POSITION_VAL)
            if (retval == HID_DEVICE_SUCCESS):
                Status2ReadSet()
                return retval
    return retval

def FocusMove(addrData):
    global focusCurrentAddr
    global focusSpeedPPS
    moveVal = abs(addrData - focusCurrentAddr)
    waitTime = WaitCalc(moveVal, focusSpeedPPS)
    retval, data = DeviceMove(DA.FOCUS_POSITION_VAL, addrData, CV.FOCUS_MASK, waitTime)
    if (retval == HID_DEVICE_SUCCESS):
        focusCurrentAddr = data
    return retval

def IrisCurrentAddrReadSet():
    global irisCurrentAddr
    retval, irisCurrentAddr = UsbCtrl.UsbRead2BytesInt(DA.IRIS_POSITION_VAL)
    return retval

def IrisParameterReadSet():
    global irisMinAddr
    global irisMaxAddr
    global irisSpeedPPS
    retval, irisMinAddr = UsbCtrl.UsbRead2BytesInt(DA.IRIS_MECH_STEP_MIN)
    if (retval != HID_DEVICE_SUCCESS):
        return retval
    retval, irisMaxAddr = UsbCtrl.UsbRead2BytesInt(DA.IRIS_MECH_STEP_MAX)
    if (retval != HID_DEVICE_SUCCESS):
        return retval
    retval, irisSpeedPPS = UsbCtrl.UsbRead2BytesInt(DA.IRIS_SPEED_VAL)
    if (retval != HID_DEVICE_SUCCESS):
        return retval 
    return retval

def IrisBacklashRead():
    retval, data = UsbCtrl.UsbRead2BytesInt(DA.IRIS_BACKLASH_CANCEL)
    return retval, data

def IrisBacklashWrite(flag):
    retval = UsbCtrl.UsbWrite(DA.IRIS_BACKLASH_CANCEL, flag)
    return retval

def IrisSpeedMinRead():
    retval, data = UsbCtrl.UsbRead2BytesInt(DA.IRIS_SPEED_MIN)
    return retval, data

def IrisSpeedMaxRead():
    retval, data = UsbCtrl.UsbRead2BytesInt(DA.IRIS_SPEED_MAX)
    return retval, data

def IrisSpeedWrite(speedPPS):
    global irisSpeedPPS
    retval = UsbCtrl.UsbWrite(DA.IRIS_SPEED_VAL, speedPPS)
    if (retval != HID_DEVICE_SUCCESS):
        return retval
    time.sleep(0.001)
    retval, irisSpeedPPS = UsbCtrl.UsbRead2BytesInt(DA.IRIS_SPEED_VAL)
    return retval

def IrisCountValRead():
    retval, count = UsbCtrl.CountRead(DA.IRIS_COUNT_VAL)
    return retval, count

def IrisCountMaxRead():
    retval, count = UsbCtrl.CountRead(DA.IRIS_COUNT_MAX)
    return retval, count

def IrisInit():
    global irisMaxAddr
    global irisMinAddr
    global irisSpeedPPS
    global irisCurrentAddr
    waitTime = WaitCalc((irisMaxAddr - irisMinAddr), irisSpeedPPS)
    retval = UsbCtrl.UsbWrite(DA.IRIS_INITIALIZE, CV.INIT_RUN_BIT)
    if (retval == HID_DEVICE_SUCCESS):
        retval = StatusWait(DA.STATUS1, CV.IRIS_MASK, waitTime)
        if (retval == HID_DEVICE_SUCCESS):
            retval, irisCurrentAddr = UsbCtrl.UsbRead2BytesInt(DA.IRIS_POSITION_VAL)
            if (retval == HID_DEVICE_SUCCESS):
                Status2ReadSet()
                return retval
    return retval

def IrisMove(addrData):
    global irisCurrentAddr
    global irisSpeedPPS
    moveVal = abs(addrData - irisCurrentAddr)
    waitTime = WaitCalc(moveVal, irisSpeedPPS)
    retval, data = DeviceMove(DA.IRIS_POSITION_VAL, addrData, CV.IRIS_MASK,waitTime)
    if (retval == HID_DEVICE_SUCCESS):
        irisCurrentAddr = data
    return retval

def OptFilterCurrentAddrReadSet():
    global optCurrentAddr
    retval, optCurrentAddr = UsbCtrl.UsbRead2BytesInt(DA.OPT_FILTER_POSITION_VAL)
    return retval

def OptFilterParameterReadSet():
    global optFilMaxAddr
    retval, optFilMaxAddr =UsbCtrl.UsbRead2BytesInt(DA.OPT_FILTER_MECH_STEP_MAX)
    if (retval != HID_DEVICE_SUCCESS):
        return retval
    OptFilterCurrentAddrReadSet()
    return retval

def OPtFilterCountValRead():
    retval, count = UsbCtrl.CountRead(DA.OPT_FILTER_COUNT_VAL)
    return retval, count

def OptFilterCountMaxRead():
    retval, count = UsbCtrl.CountRead(DA.OPT_FILTER_COUNT_MAX)
    return retval, count

def OptFilterInit():
    global optFilMaxAddr
    global optCurrentAddr
    waitTime = WaitCalc((optFilMaxAddr + 1), CV.OPT_FILTER_SPEED)
    retval = UsbCtrl.UsbWrite(DA.OPT_FILTER_INITIALIZE, CV.INIT_RUN_BIT)
    if (retval == HID_DEVICE_SUCCESS):
        retval  = StatusWait(DA.STATUS1, CV.OPT_FILTER_MASK, waitTime)
        if (retval == HID_DEVICE_SUCCESS):
            retval, optCurrentAddr = UsbCtrl.UsbRead2BytesInt(DA.OPT_FILTER_POSITION_VAL) 
            if (retval == HID_DEVICE_SUCCESS):
                Status2ReadSet()
                return retval
        return retval

def OptFilterMove(addrData):
    global optCurrentAddr
    moveVal = abs (addrData - optCurrentAddr)
    waitTime = WaitCalc(moveVal, CV.OPT_FILTER_SPEED)
    retval, data = DeviceMove(DA.OPT_FILTER_POSITION_VAL, addrData, waitTime)
    if (retval == HID_DEVICE_SUCCESS):
        optCurrentAddr = data
        return retval

def DeviceMove(segmentOffset, addrData, mask, waitTime):
    retval = UsbCtrl.UsbWrite(segmentOffset, addrData)
    if (retval == HID_DEVICE_SUCCESS):
        retval = StatusWait(DA.STATUS1, mask, waitTime)
        if (retval == HID_DEVICE_SUCCESS):
            retval, data = UsbCtrl.UsbRead2BytesInt(segmentOffset)
            if (retval != HID_DEVICE_SUCCESS):
                return retval, addrData
            addrData = data
    return retval, addrData

def StatusWait(segmentOffset, statusMask, waitTime):
    tmp = 0
    readStatus = 0
    while ((readStatus & statusMask) != statusMask):
        retval, data = UsbCtrl.UsbRead2BytesInt(segmentOffset)
        if (retval != HID_DEVICE_SUCCESS):
            return retval
        readStatus = data
        tmp += 1
        if (tmp >= CV.LOW_HIGH_WAIT):
            return CV.LOWHI_ERROR
    tmp = 0
    readStatus = 0xFF
    while ((readStatus & statusMask) != 0):
        retval, data = UsbCtrl.UsbRead2BytesInt(segmentOffset)
        if (retval != HID_DEVICE_SUCCESS):
            return retval
        readStatus = data
        tmp += 1
        if (tmp >= waitTime):
            return CV.HILOW_ERROR
        time.sleep(0.001)
    return retval
